File: sdk/python/src/waeasi/wasi/clocks.py
"""waeasi.wasi.clocks — monotonic + wall-clock + sleep."""

from __future__ import annotations

import asyncio
import time
from typing import Callable, Optional

_MONO_BASE_NS: Optional[int] = None
_MONO_OVERRIDE: Optional[Callable[[], int]] = None
_WALL_OVERRIDE: Optional[Callable[[], int]] = None


def bind_host(
    *,
    mono_now: Optional[Callable[[], int]] = None,
    wall_now: Optional[Callable[[], int]] = None,
) -> None:
    global _MONO_OVERRIDE, _WALL_OVERRIDE
    _MONO_OVERRIDE = mono_now
    _WALL_OVERRIDE = wall_now


def monotonic_now() -> int:
    """Nanoseconds since instance start.  Strictly monotonic."""
    if _MONO_OVERRIDE is not None:
        return _MONO_OVERRIDE()
    global _MONO_BASE_NS
    cur = time.monotonic_ns()
    if _MONO_BASE_NS is None:
        _MONO_BASE_NS = cur
    return cur - _MONO_BASE_NS


def wall_now() -> int:
    """Unix-epoch nanoseconds."""
    if _WALL_OVERRIDE is not None:
        return _WALL_OVERRIDE()
    return time.time_ns()


async def sleep(ms: float) -> None:
    """Cooperative sleep — yields to the asyncio loop."""
    if ms <= 0:
        await asyncio.sleep(0)
        return
    await asyncio.sleep(ms / 1000.0)

File: sdk/typescript/src/wasi/clocks.ts
// @waeasi/sdk — wasi:clocks facade
//
// Two clocks per WASI 0.2:
//   * monotonic-clock — never goes backwards, ns resolution, base is
//     instance-local (NOT the kernel's global mono).
//   * wall-clock — unix-epoch ns; may jump under NTP.
//
// At componentize time these stubs are linked directly to the host
// imports.  In dev-mode we fall back to performance.now() / Date.now().

let MONO_BASE_NS: bigint | null = null;
let HOST: { monoNow?: () => bigint; wallNow?: () => bigint } = {};

export function bindHost(h: { monoNow?: () => bigint; wallNow?: () => bigint }): void {
    HOST = h;
}

/** Nanoseconds since instance start.  Strictly monotonic. */
export function monotonicNow(): bigint {
    if (HOST.monoNow) return HOST.monoNow();
    if (typeof performance !== "undefined") {
        if (MONO_BASE_NS === null) {
            MONO_BASE_NS = BigInt(Math.trunc(performance.now() * 1e6));
        }
        const cur = BigInt(Math.trunc(performance.now() * 1e6));
        return cur - MONO_BASE_NS;
    }
    // last-resort: Date as monotonic proxy (dev only)
    if (MONO_BASE_NS === null) MONO_BASE_NS = BigInt(Date.now()) * 1_000_000n;
    return BigInt(Date.now()) * 1_000_000n - MONO_BASE_NS;
}

/** Unix epoch nanoseconds. */
export function wallNow(): bigint {
    if (HOST.wallNow) return HOST.wallNow();
    return BigInt(Date.now()) * 1_000_000n;
}

/** High-level helper: sleep `ms` milliseconds (cooperative). */
export function sleep(ms: number): Promise<void> {
    return new Promise<void>((resolve) => {
        const start = monotonicNow();
        const target = start + BigInt(Math.trunc(ms * 1e6));
        const tick = (): void => {
            if (monotonicNow() >= target) resolve();
            else setTimeout(tick, Math.max(1, Math.min(16, ms)));
        };
        tick();
    });
}
